File: leetcode/python/33_search_in_rotated_sorted_array.py
"""
T: O(logN)
S: O(logN)

The one-pass solution to the problem. It can be done through observing the
invariances. We start the usual binary search, and if we found the appropriate
number we just return the index. Otherwise, there are two cases to consider:
1) nums[lo] <= nums[mid], part to the left of the mid is sorted, and if the
target can be in the sorted part, we go there, otherwise, we go right
2) otherwise the right part is sorted, and if the target is possibly in it,
we go right, otherwise, we go left
"""

from typing import List


class Solution:

    def search(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if nums[mid] == target:
                return mid
            if nums[lo] <= nums[mid]:
                if nums[lo] <= target < nums[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1
            else:
                if nums[mid] < target <= nums[-1]:
                    lo = mid + 1
                else:
                    hi = mid - 1
        return -1

File: leetcode/java/33SearchInRotatedSortedArray.java
/**
 * T: O(logN) S: O(1)
 *
 * <p>First, find the index of the smallest element. Now we can can find which part of the array
 * target belongs to and execute binary search.
 */
class Solution {
  private final int NUM_NOT_FOUND = -1;

  public int search(int[] nums, int target) {
    if (nums.length == 0) {
      return NUM_NOT_FOUND;
    }
    int minIndex = findMinIndex(nums);
    if (target >= nums[minIndex] && target <= nums[nums.length - 1]) {
      return binarySearch(nums, target, minIndex, nums.length - 1);
    }
    return binarySearch(nums, target, 0, minIndex - 1);
  }

  private int findMinIndex(int[] nums) {
    int left = 0;
    int right = nums.length - 1;
    while (left != right) {
      int mid = (left + right) / 2;
      if (nums[mid] > nums[right]) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }
    return left;
  }

  private int binarySearch(int[] nums, int target, int left, int right) {
    while (left <= right) {
      int mid = (left + right) / 2;
      if (nums[mid] > target) {
        right = mid - 1;
      } else if (nums[mid] < target) {
        left = mid + 1;
      } else {
        return mid;
      }
    }
    return NUM_NOT_FOUND;
  }
}
